language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -40,7 +40,8 @@ newtype DynamicBuiltinNameTypes = DynamicBuiltinNameTypes
 -- | Configuration of the type checker.
 data TypeConfig = TypeConfig
     { _typeConfigNormalize           :: Bool
-      -- ^ Whether to normalize type annotations
+      -- ^ Whether to normalize type annotations. When set to 'True', the
+      -- typechecker is set to unlimited mode and will not run out of gas.
```

Comment:
IMO these are two separate things. One determines whether or not you do normalization, the other one determines how many steps you take. You can plausibly pick the various different combinations depending on what you want to do. I think we should instead make the amount of gas a `Maybe` or something.

Reply:
Using a `Maybe Natural` for gas is a good idea, but we actually do want these connected. See https://iohk.myjetbrains.com/youtrack/issue/CGP-276

Although looking back it looks like we want to bound how long *normalization* takes which is slightly different. I think this PR needs a bit more work.

Reply:
I'd be inclined to have them be separate at the library level, although we might not expose that to users. It just makes the implementation more obvious IMO.

If we're going to change the gas model to the one proposed in that ticket (normalization length), can we do that in a separate PR?

Reply:
Alright, this should be ready for inspection.